From 6d9c34580ba0438614544124d383c8867690388c Mon Sep 17 00:00:00 2001
From: "xueqiang.li" <xueqiang.li@amlogic.com>
Date: Tue, 11 Feb 2014 15:45:18 +0800
Subject: [PATCH 3258/5965] PD #87056: wifi power enable GPIOAO_6 and GPIOX_11

commit a5a0a0d46eb82fa981103c67c35890a722d944d6
Author: xueqiang.li <xueqiang.li@amlogic.com>
Date:   Tue Feb 11 14:59:26 2014 +0800

    change variable name to wifi_power_on_pin2 and wifi_power_gpio2

commit 7ffb395fd86846c84e4720c1a3a71f71569b5f2a
Author: xueqiang.li <xueqiang.li@amlogic.com>
Date:   Tue Feb 11 10:00:37 2014 +0800

    support WIFI enabled by GPIOAO_6 or GPIOX_11

commit 0b0eab6ba09e6d12977aaacf295813b2dfcc5b2a
Author: xueqiang.li <xueqiang.li@amlogic.com>
Date:   Mon Feb 10 17:28:21 2014 +0800

    support WIFI enabled by GPIOAO_6 or GPIOX_11
---
 arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd  |  1 +
 arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd  |  1 +
 arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd  |  1 +
 .../arm/boot/dts/amlogic/meson8_k101_512M.dtd |  1 +
 arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd  |  2 +
 arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd  |  2 +
 arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd |  2 +
 arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd |  2 +
 arch/arm/plat-meson/include/plat/wifi_power.h |  1 +
 drivers/amlogic/wifi/wifi_dt.c                | 44 ++++++++++++++++++-
 drivers/amlogic/wifi/wifi_power.c             | 32 ++++++++++++++
 11 files changed, 87 insertions(+), 2 deletions(-)

diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
index 55939112e9f6..56f6f61d7bd4 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
@@ -387,6 +387,7 @@ void root_func(){
 		irq_num = <4>;
 		irq_trigger_type = "GPIO_IRQ_HIGH";
 		power_on_pin = "GPIOAO_6";
+		power_on_pin2 = "GPIOX_11";
 		clock_32k_pin = "GPIOX_10";
 	};
 
diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd
index c4daf71ef14d..4582371d4d25 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd
@@ -386,6 +386,7 @@ void root_func(){
 		irq_num = <4>;
 		irq_trigger_type = "GPIO_IRQ_HIGH";
 		power_on_pin = "GPIOAO_6";
+		power_on_pin2 = "GPIOX_11";
 		clock_32k_pin = "GPIOX_10";
 	};
 
diff --git a/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd
index 0e7ca06ec77f..89a27e0d6a70 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd
@@ -1666,6 +1666,7 @@ lcd_LD070WX4:lcd_LD070WX4{
             dev_name = "wifi_power";
             status = "okay";
             power_gpio = "GPIOAO_6";
+            power_gpio2 = "GPIOX_11";
 	};  
     
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd b/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd
index c4e919da7c7a..eef6a1a00fcf 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd
@@ -1613,6 +1613,7 @@ lcd_CLAA101FP05XG:lcd_CLAA101FP05XG{	//k101
             dev_name = "wifi_power";
             status = "okay";
             power_gpio = "GPIOAO_6";
+            power_gpio2 = "GPIOX_11";
 	};  
     
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
index 169b4f94e682..26d8d61594d7 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
@@ -361,6 +361,7 @@ void root_func(){
 		irq_num = <4>;
 		irq_trigger_type = "GPIO_IRQ_HIGH";
 		power_on_pin = "GPIOAO_6";
+		power_on_pin2 = "GPIOX_11";
 		clock_32k_pin = "GPIOX_10";
 	};
 	
@@ -372,6 +373,7 @@ void root_func(){
             dev_name = "wifi_power";
             status = "okay";
             power_gpio = "GPIOAO_6";
+            power_gpio2 = "GPIOX_11";
 	}; 
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
index 87a7a3349344..fe336db57d79 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
@@ -361,6 +361,7 @@ void root_func(){
 		irq_num = <4>;
 		irq_trigger_type = "GPIO_IRQ_HIGH";
 		power_on_pin = "GPIOAO_6";
+		power_on_pin2 = "GPIOX_11";
 		clock_32k_pin = "GPIOX_10";
 	};
 	
@@ -372,6 +373,7 @@ void root_func(){
             dev_name = "wifi_power";
             status = "okay";
             power_gpio = "GPIOAO_6";
+            power_gpio2 = "GPIOX_11";
 	}; 
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd
index 43333a423b33..8fad867320a3 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd
@@ -361,6 +361,7 @@ void root_func(){
 		irq_num = <4>;
 		irq_trigger_type = "GPIO_IRQ_HIGH";
 		power_on_pin = "GPIOAO_6";
+		power_on_pin2 = "GPIOX_11";
 		clock_32k_pin = "GPIOX_10";
 	};
 	
@@ -372,6 +373,7 @@ void root_func(){
             dev_name = "wifi_power";
             status = "okay";
             power_gpio = "GPIOAO_6";
+            power_gpio2 = "GPIOX_11";
 	}; 
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd
index 53c9ec41ce65..fdc95d80c34c 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd
@@ -360,6 +360,7 @@ void root_func(){
 		irq_num = <4>;
 		irq_trigger_type = "GPIO_IRQ_HIGH";
 		power_on_pin = "GPIOAO_6";
+		power_on_pin2 = "GPIOX_11";
 		clock_32k_pin = "GPIOX_10";
 	};
 	
@@ -371,6 +372,7 @@ void root_func(){
             dev_name = "wifi_power";
             status = "okay";
             power_gpio = "GPIOAO_6";
+            power_gpio2 = "GPIOX_11";
 	}; 
 
 /// ***************************************************************************************
diff --git a/arch/arm/plat-meson/include/plat/wifi_power.h b/arch/arm/plat-meson/include/plat/wifi_power.h
index 1df52c69ff13..2d4b69fa58cf 100644
--- a/arch/arm/plat-meson/include/plat/wifi_power.h
+++ b/arch/arm/plat-meson/include/plat/wifi_power.h
@@ -2,6 +2,7 @@
 #define _WIFI_POWER_H_
 struct wifi_power_platform_data{
     int power_gpio;
+    int power_gpio2;
     int (*set_power)(int val);
     int (*set_reset)(int val);
     int (*set_carddetect)(int val);
diff --git a/drivers/amlogic/wifi/wifi_dt.c b/drivers/amlogic/wifi/wifi_dt.c
index 110f93735e4d..283c99dbf220 100755
--- a/drivers/amlogic/wifi/wifi_dt.c
+++ b/drivers/amlogic/wifi/wifi_dt.c
@@ -12,12 +12,15 @@
 
 #define OWNER_NAME "sdio_wifi"
 
+int wifi_power_gpio2 = 0;
+
 struct wifi_plat_info {
 	int interrupt_pin;
 	int irq_num;
 	int irq_trigger_type;
 
 	int power_on_pin;
+	int power_on_pin2;
 
 	int clock_32k_pin;
 
@@ -107,6 +110,16 @@ static int wifi_dev_probe(struct platform_device *pdev)
 		CHECK_PROP(ret, "power_on_pin", value);
 		plat->power_on_pin = amlogic_gpio_name_map_num(value);
 
+		if(!(ret = of_property_read_string(pdev->dev.of_node, "power_on_pin2", &value)))
+			wifi_power_gpio2 = 1;
+		
+		if(wifi_power_gpio2){
+			plat->power_on_pin2 = amlogic_gpio_name_map_num(value);
+		}
+		else{
+			printk("wifi_dev_probe : there is no wifi_power_gpio2 setup in DTS file!\n");
+		}
+
 		ret = of_property_read_string(pdev->dev.of_node, "clock_32k_pin", &value);
 		CHECK_PROP(ret, "clock_32k_pin", value);
 		plat->clock_32k_pin = amlogic_gpio_name_map_num(value);
@@ -114,9 +127,11 @@ static int wifi_dev_probe(struct platform_device *pdev)
 		plat->plat_info_valid = 1;
 		
 		printk("interrupt_pin=%d, irq_num=%d, irq_trigger_type=%d, "
-				"power_on_pin=%d, clock_32k_pin=%d\n", 
+				"power_on_pin=%d,"
+				"clock_32k_pin=%d\n", 
 				plat->interrupt_pin, plat->irq_num, plat->irq_trigger_type, 
-				plat->power_on_pin, plat->clock_32k_pin);
+				plat->power_on_pin,
+				plat->clock_32k_pin);
 	}
 #endif
 
@@ -207,6 +222,15 @@ int wifi_setup_dt()
 	CHECK_RET(ret);
 	SHOW_PIN_OWN("power_on_pin", wifi_info.power_on_pin);
 
+	if(wifi_power_gpio2){
+		SHOW_PIN_OWN("power_on_pin2", wifi_info.power_on_pin2);
+		ret = amlogic_gpio_request(wifi_info.power_on_pin2, OWNER_NAME);
+		CHECK_RET(ret);
+		ret = amlogic_gpio_direction_output(wifi_info.power_on_pin2, 0, OWNER_NAME);
+		CHECK_RET(ret);
+		SHOW_PIN_OWN("power_on_pin2", wifi_info.power_on_pin2);
+	}
+
 	return 0;
 }
 EXPORT_SYMBOL(wifi_setup_dt);
@@ -223,6 +247,12 @@ void wifi_teardown_dt()
 	
 	ret = amlogic_gpio_free(wifi_info.power_on_pin, OWNER_NAME);
 	CHECK_RET(ret);
+	
+	if(wifi_power_gpio2)
+	{
+		ret = amlogic_gpio_free(wifi_info.power_on_pin2, OWNER_NAME);
+		CHECK_RET(ret);
+	}
 	ret = amlogic_gpio_free(wifi_info.interrupt_pin, OWNER_NAME);
 	CHECK_RET(ret);
 	wifi_request_32k_clk(0, OWNER_NAME);
@@ -284,10 +314,20 @@ void extern_wifi_set_enable(int is_on)
 	if (is_on) {
 		ret = amlogic_gpio_direction_output(wifi_info.power_on_pin, 1, OWNER_NAME);
 		CHECK_RET(ret);
+		if(wifi_power_gpio2){
+			ret = amlogic_gpio_direction_output(wifi_info.power_on_pin2, 1, OWNER_NAME);
+			CHECK_RET(ret);
+		}
 		printk("WIFI  Enable! %d\n", wifi_info.power_on_pin);
 	} else {
 		ret = amlogic_gpio_direction_output(wifi_info.power_on_pin, 0, OWNER_NAME);
 		CHECK_RET(ret);
+		
+		if(wifi_power_gpio2){
+			ret = amlogic_gpio_direction_output(wifi_info.power_on_pin2, 0, OWNER_NAME);
+			CHECK_RET(ret);
+		}
+		
 		printk("WIFI  Disable! %d\n", wifi_info.power_on_pin);
 	
 	}
diff --git a/drivers/amlogic/wifi/wifi_power.c b/drivers/amlogic/wifi/wifi_power.c
index 924220fb72ca..d2e8a7ffa57c 100755
--- a/drivers/amlogic/wifi/wifi_power.c
+++ b/drivers/amlogic/wifi/wifi_power.c
@@ -31,6 +31,7 @@ static dev_t wifi_power_devno;
 static struct cdev *wifi_power_cdev = NULL;
 static struct device *devp=NULL;
 struct wifi_power_platform_data *pdata = NULL;
+int wifi_power_on_pin2 = 0;
     
 static int wifi_power_probe(struct platform_device *pdev);
 static int wifi_power_remove(struct platform_device *pdev);
@@ -92,6 +93,9 @@ static int wifi_power_ioctl(struct file *filp, unsigned int cmd, unsigned long a
     }
     
     amlogic_gpio_request(pdata->power_gpio,WIFI_POWER_MODULE_NAME);
+
+	if(wifi_power_on_pin2)
+ 	   amlogic_gpio_request(pdata->power_gpio2,WIFI_POWER_MODULE_NAME);
     
 	switch (cmd) 
 	{
@@ -281,6 +285,16 @@ static void usb_wifi_power(int is_power)
 	    else
 	        amlogic_gpio_direction_output(pdata->power_gpio, 1, WIFI_POWER_MODULE_NAME); 
     }
+	if(wifi_power_on_pin2){
+	    	if(pdata->power_gpio2 > 0)
+		{
+		    if(is_power)
+	            amlogic_gpio_direction_output(pdata->power_gpio2, 0, WIFI_POWER_MODULE_NAME); 	  
+		    else
+		        amlogic_gpio_direction_output(pdata->power_gpio2, 1, WIFI_POWER_MODULE_NAME); 
+	    }
+	}
+	
 #else    
     CLEAR_CBUS_REG_MASK(PREG_PAD_GPIO6_EN_N, (1<<11));
     if (is_power)//is_power
@@ -317,6 +331,24 @@ static int wifi_power_probe(struct platform_device *pdev)
 	        //mcli pdata->usb_set_power(0);    //power on   
 	        //pdata->usb_set_power(1);    //power on   
 	     }
+		 
+	   if(!(ret = of_property_read_string(pdev->dev.of_node, "power_gpio2", &str)))
+			wifi_power_on_pin2 = 1;
+	   else{
+		printk("wifi_dev_probe : there is no wifi_power_on_pin2 setup in DTS file!\n");
+	   }
+	   
+	   if(wifi_power_on_pin2){
+		    if(ret)
+		     {  
+		        printk("Error: can not get power_gpio2 name------%s %d\n",__func__,__LINE__);
+		        return -1;
+		     }else{
+			 pdata->power_gpio2 = amlogic_gpio_name_map_num(str);
+		        printk("wifi_power power_gpio2 is %d\n",pdata->power_gpio2);
+		     }
+	   }
+		 
 	}
     pdev->dev.platform_data = pdata;
     
-- 
2.19.0

